drivers: charger: add initial support for the charger module found in AXP2101 MFD device

valeriosetti opened the pull request:
This PR adds initial support for the charger module which is part of the AXP2101 MFD device. The same IC also implements regulators, but those are already supported in Zephyr.

Finally the DTS of the lilygo/twatch_s3 board is also extended to enable this driver.

Depends on #88053 in order to have CI fully green

valeriosetti committed:
drivers: charger: add charger support in X-Powers AXP2101

AXP2101 is MFD device. Zephyr already support the regulator part. This
commit introduces intial support for the charger one.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
boards: twatch_s3: enable battery charger

Enable the battery charger module form the X-Powers AXP2101 IC.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

rriveramcrus commented on boards/lilygo/twatch_s3/twatch_s3_procpu.dts:
```diff
@@ -139,6 +139,15 @@
+
+		charger {
+			compatible = "x-powers,axp2101-charger";
+			constant-charge-current-ma = <300>;
+			constant-charge-voltage-mv = <4200>;
+			termination-current-ma = <125>;
```
These all exist as standard properties in `dts/bindings/battery/battery.yaml` so we should use them if possible.

rriveramcrus commented on drivers/charger/Kconfig.axp2101:
```diff
@@ -0,0 +1,23 @@
+
+config CHARGER_AXP2101_INIT_PRIORITY
+	int "AXP2101 charger driver init priority"
+	default 86
```
We could rely on `CHARGER_INIT_PRIORITY` as a default since it is greater than the `MFD_INIT_PRIORITY` default value.

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+
+#include <errno.h>
+
+#include "zephyr/device.h"
+#include "zephyr/drivers/charger.h"
+#include "zephyr/drivers/i2c.h"
+#include "zephyr/kernel.h"
+#include "zephyr/sys/util.h"
+#include "zephyr/logging/log.h"
```
These should be absolute `#include`s rather than relatives. `#include <foo.h>` rather than `#include "foo.h"`

rriveramcrus commented on drivers/charger/charger_axp2101.h:
```diff
@@ -0,0 +1,91 @@
+#define VHTF_CHG_SETTING			0x55
+#define VLTF_WORK_SETTING			0x56
+#define VHTF_WORK_SETTING			0x57
+#define JIETA_ENABLE_CONTROL			0x58
+#define JIETA_SETTING_START			0x59
+#define JIETA_SETTING_END			0x5B
```
https://home.jeita.or.jp/upload_file/20210407112651_JbDgUSWGiB.pdf
Japan Electronics and Information Technology Industries Association

rriveramcrus replied:
The implication here is that JIETA_* should be JEITA_*

valeriosetti replied:
My plan was indeed to ask you about this comment, but then I forgot to do it. Thanks for the clarification!

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+{
+	const struct axp2101_config *const config = dev->config;
+	struct axp2101_data *data = dev->data;
+	uint32_t tmp = U2M(val->const_charge_current_ua);
```
reverse pyramid declarations where possible please.

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+static int get_pre_charge_current_ua(const struct device *dev, union charger_propval *val)
+{
+	const struct axp2101_config *const config = dev->config;
+	uint8_t tmp = 0;
```
doesn't seem necessary to initialize `tmp` to 0

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+				      union charger_propval *val)
+{
+	const struct axp2101_config *const config = dev->config;
+	uint8_t tmp = 0;
```
doesn't seem necessary to initialize `tmp` to 0

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+		return ret;
+	}
+
+	if (!IS_BIT_SET(tmp, 4)) {
+		val->charge_term_current_ua = 0;
+	} else {
+		tmp = tmp & GENMASK(3, 0);
+		val->charge_term_current_ua = M2U(25 * ((uint32_t) val));
+	}
```
Could you define these masks? I guess one is the termination current threshold bitfield, but idk. I also guess the fourth bit is the enable bit for the termination current circuitry, but again idk.

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+		return ret;
+	}
+
+	if (tmp > 5) {
```
I'd prefer a `#define` or `ARRAY_SIZE(constant_charge_voltage_lut)` instead of 5.

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+	unsigned int cc_current_ma;
+	unsigned int cc_voltage_mv;
+	unsigned int termination_current_ma;
+	bool vbackup_enable;
```
This is not configurable at runtime and is platform dependent, I'd suggest moving it to `axp2101_config`.

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+	if (ret) {
+		return ret;
+	}
+	val->online = (tmp == 1) ? CHARGER_ONLINE_FIXED : CHARGER_ONLINE_OFFLINE;
```
Is bit 0 AC_OK? Is there anything else in this register?

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,394 @@
+		return ret;
+	}
+
+	LOG_INF("AXP2101 initialized");
```
I don't feel this is necessary. It also feels somewhat misleading as this is an MFD and the regulator init may have failed.

valeriosetti commented on drivers/charger/charger_axp2101.h:
```diff
@@ -0,0 +1,91 @@
+#define _CHARGER_AXP2101_H_
+
+#define PMU_STATUS1				0x00
+	#define VBUS_GOOD_INDICATION		BIT(5)
```
TBH I don't know. I like this pattern because it visually helps readers in understanding that these lines are the bits of the parent register. But I don´t know if there's any standardized way of doing it

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+};
+
+struct axp2101_data {
+	unsigned int cc_current_ua;
+	unsigned int cc_voltage_uv;
+	unsigned int termination_current_ua;
```
Might be safer if these were `uint32_t` as the respective members of `charger_propval` are `uint32_t`

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+		return ret;
+	}
+
+	if (!IS_BIT_SET(tmp, 4)) {
```
use `CHARGE_TERMINATION_ENABLE` instead of 4

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+	int ret;
+
+	if (config->vbackup_enable) {
+		ret = i2c_reg_write_byte_dt(&config->i2c, CHARGER_CONTROL,
```
This is clobbering some of the fuel gauge control bits. If we ever add fuel gauge support (I'd love to see it), this will become a problem. Let's use `i2c_reg_update_byte_dt` instead.

valeriosetti replied:
> I'd love to see it

I'm planning to add it after the charging part, but I need first to understand how it works (probably by looking at the sample code provided for the Lilygo T-Watch S3 board I'm using to develop this driver)

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+	}
+
+	tmp = tmp & CHARGING_STATUS;
+	if ((tmp == 0) || (tmp == 1)) {
+		val->charge_type = CHARGER_CHARGE_TYPE_TRICKLE;
+	} else if ((tmp >= 2) && (tmp <= 4)) {
+		val->charge_type = CHARGER_CHARGE_TYPE_STANDARD;
+	} else if (tmp == 5) {
+		val->charge_type = CHARGER_CHARGE_TYPE_UNKNOWN;
```
I would appreciate some `#define`s for these magic numbers

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+	}
+
+	tmp = tmp & CHARGING_STATUS;
+	if ((tmp >= 0) && (tmp <= 3)) {
+		val->status = CHARGER_STATUS_CHARGING;
+	} else if (tmp == 4) {
+		val->status = CHARGER_STATUS_FULL;
+	} else if (tmp == 5) {
```
Same with the magic numbers here

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+	if (ret) {
+		return ret;
+	}
+	val->precharge_current_ua = 25000 * ((uint32_t) tmp);
```
magic number needs a definition

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+		val->charge_term_current_ua = 0;
+	} else {
+		tmp = tmp & TERMINATION_CURRENT_LIMIT;
+		val->charge_term_current_ua = 25000 * ((uint32_t) tmp);
```
magic number needs definition

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+	if ((tmp > 200000) || ((tmp % 25000) != 0)) {
+		return -EINVAL;
+	}
+	tmp = (uint8_t) (tmp / 25000);
```
more magic number cleanup

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+	uint32_t mask, tmp;
+
+	tmp = val->charge_term_current_ua;
+	if ((tmp > 200000) || ((tmp % 25000) != 0)) {
```
what if `tmp` is 0, is that a valid value? Semantically, that is functionally equivalent to disabling the termination current circuitry so you may want to clear `CHARGE_TERMINATION_ENABLE` or reject 0 as a valid value.

valeriosetti replied:
According to the AXP2101 datasheet you can enable termination current and set it to 0 mA, but you are right it doesn't make much sense. I will disable it in this case. Thanks

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,373 @@
+#define AXP2101_INIT(inst)                                                                         \
+	static const struct axp2101_config axp2101_config_##inst = {                               \
+		.i2c = I2C_DT_SPEC_GET(DT_PARENT(DT_INST(inst, DT_DRV_COMPAT))),                   \
+		.vbackup_enable = DT_INST_PROP_OR(inst, vbackup_enable, false),                    \
```
`bool` is initialized to `false` already, so we do not need `DT_INST_PROP_OR`; instead `DT_INST_PROP` will suffice.

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,381 @@
+		ret = i2c_reg_update_byte_dt(&config->i2c, CHARGER_CONTROL,
+					     BUTTON_BATTERY_CHARGE_ENABLE,
+					     BUTTON_BATTERY_CHARGE_ENABLE);
+		if (ret) {
```
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html

As-is this violates [Rule 14.4](https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_14_04.c), how about `if (ret < 0)` instead? That way the expression resolves to a boolean.

There are a few other lines in the driver with this same issue.

rriveramcrus commented:
Overall this is looking really good, once the MISRA-C 2012 related fixes are in place I feel ready to approve. 

https://github.com/zephyrproject-rtos/zephyr/pull/86791/commits/8d321c8fca3551011f1abf5870cf7a87b872ab79#diff-db5eebc0f089960f1220673dde2c861af4e2b36adca08183704729335562c2b5
A bonus item to address, if you have the time, would be to add an entry to `tests/drivers/build_all/charger/i2c.dtsi` as was done in this commit.

valeriosetti committed:
tests: drivers: build_all: add support for "x-powers,axp2101-charger"

Extend the *.dtsi file to include also an "x-powers,axp2101-charger"
instance.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

rriveramcrus commented on tests/drivers/build_all/charger/i2c.dtsi:
```diff
@@ -46,3 +46,15 @@ sbs-charger@3 {
 };
+
+axp2101@4 {
+		compatible = "x-powers,axp2101";
+		reg = <0x34>;
+
+		charger {
+			compatible = "x-powers,axp2101-charger";
+			constant-charge-current-max-microamp = <300000>;
+			constant-charge-voltage-max-microvolt = <4200000>;
+			charge-term-current-microamp = <100000>;
+		};
```
This is one indent level too far

valeriosetti replied:
Ops, sorry for that ':)

rriveramcrus reviewed (dismissed):
Thank you for your contribution!

soburi commented on drivers/charger/Kconfig.axp2101:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2025 BayLibre SAS
+# SPDX-License-Identifier: Apache-2.0
+
+config CHARGER_AXP2101
```
Other AXP2101 function share the implementation with AXP192.

https://github.com/zephyrproject-rtos/zephyr/blob/6c989dee10c5795bb85dc59ad87d8898ee7940d0/drivers/regulator/Kconfig.axp192#L4-L8

Can the charger feature do as so?

soburi replied:
https://github.com/m5stack/M5-Schematic/blob/master/Core/AXP192%20Datasheet_v1.1_en_draft_2211.pdf

The Chinese version seems to be a slightly newer revision, but this one should be fine.

valeriosetti replied:
Thanks a lot for the link!
I checked the first 2 registers just as example (i.e. those at address 0x00 and 0x01) and although the purpose look similar, they differ in bits location and meaning. I can make this driver handle both, but it would look like a merge of 2 separate drivers plus some glue to accommodate them in the same file. IMO it's worth having 2 separate source files, one for axp2101 and one for axp192. Wdyt?

rriveramcrus replied:
@soburi and @valeriosetti 
Given the bit fields don't match and this charging IC seems to be targeting more resource constrained systems such as the smart watch, I would vote to keep the source files separate.

soburi replied:
It certainly seems like there are differences at the design level.
Let's leave it.

soburi commented on drivers/charger/charger_axp2101.h:
```diff
@@ -0,0 +1,100 @@
+	#define CHARGE_DONE			0x4
+	#define NOT_CHARGING			0x5
+
+#define DATA_BUFFER_START			0x04
```
The AXP2101 is configured as an MFD, and each function defines only the necessary registers.

https://github.com/zephyrproject-rtos/zephyr/blob/6c989dee10c5795bb85dc59ad87d8898ee7940d0/drivers/regulator/regulator_axp192.c#L34-L53

Can you do something like that?

soburi replied:
One more.
If it's not referenced anywhere else, it's better to put it in .c.

valeriosetti replied:
Done ;)

rriveramcrus commented on drivers/charger/charger_axp2101.c:
```diff
@@ -0,0 +1,382 @@
+
+LOG_MODULE_REGISTER(charger_axp2101, CONFIG_CHARGER_LOG_LEVEL);
+
+#define PRE_CHARGE_CURRENT_STEP		25000 /* uA */
+#define TERMINATION_CURRENT_STEP	25000 /* uA */
```
```c
#define PRE_CHARGE_CURRENT_STEP_UA
#define TERMINATION_CURRENT_STEP_UA
```
Typical convention is to suffix the units like so

rriveramcrus commented:
This does need a rebase now that #86791 was merged

rriveramcrus commented on tests/drivers/build_all/charger/i2c.dtsi:
```diff
@@ -54,3 +54,15 @@ bq257130@4 {
 	system-voltage-min-threshold-microvolt = <3328000>;
 };
+
+axp2101@4 {
```
`axp2101@5 {`
Should be five instead four since the BQ25713 claims four. The five is just a placeholder value, it does not have any real-world meaning.

rriveramcrus commented on tests/drivers/build_all/charger/i2c.dtsi:
```diff
@@ -54,3 +54,15 @@ bq257130@4 {
+
+axp2101@4 {
+	compatible = "x-powers,axp2101";
+	reg = <0x34>;
```
Should also be five here.

valeriosetti replied:
Yes, sorry, I just did a quick rebase, but I didn't carefully check the result. It should be fixed now

rriveramcrus reviewed (approved):
Thank you for your contribution!
